compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -0,0 +1,537 @@
+  return tilingResult;
+}
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// SwapExtractSliceWithTiledProducer
+//===----------------------------------------------------------------------===//
+
+/// Pattern to swap a `tilinginterface op` -> `tensor.extract_slice` with
+/// `tensor.extract_slice` of operands of the op -> tiled `tilinginterface op`.
+struct SwapExtractSliceWithTiledProducer
+    : public OpRewritePattern<tensor::ExtractSliceOp> {
+  using OpRewritePattern<tensor::ExtractSliceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExtractSliceOp sliceOp,
+                                PatternRewriter &rewriter) const override {
+    OpResult producer = sliceOp.source().dyn_cast<OpResult>();
+    if (!producer) {
+      return rewriter.notifyMatchFailure(sliceOp, "source uses bb arg");
+    }
+    FailureOr<Value> tiledProducer =
+        tensor::replaceExtractSliceWithTiledProducer(rewriter, sliceOp,
+                                                     producer);
+    if (failed(tiledProducer)) {
+      return failure();
+    }
+    rewriter.replaceOp(sliceOp, tiledProducer.getValue());
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// SwapExtractSliceWithDispatchTensorLoad
+//===----------------------------------------------------------------------===//
+
+/// Pattern to swap `flow.dispatch.tensor.load` -> `tensor.extract_slice` with
+/// `flow.dispatch.tensor.load` of the slice.
+struct SwapExtractSliceWithDispatchTensorLoad
+    : public OpRewritePattern<tensor::ExtractSliceOp> {
+  using OpRewritePattern<tensor::ExtractSliceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExtractSliceOp sliceOp,
+                                PatternRewriter &rewriter) const override {
+    auto loadOp =
+        sliceOp.source().getDefiningOp<IREE::Flow::DispatchTensorLoadOp>();
+    if (!loadOp) return failure();
+
+    SmallVector<OpFoldResult> combinedOffsets, combinedSizes, combinedStrides;
+    if (failed(IREE::Flow::foldOffsetsSizesAndStrides(
+            rewriter, loadOp.getLoc(), loadOp, sliceOp, loadOp.getDroppedDims(),
+            combinedOffsets, combinedSizes, combinedStrides))) {
+      return rewriter.notifyMatchFailure(
+          sliceOp, "failed to fold offsets, sizes and strides with load op");
+    }
+    rewriter.replaceOpWithNewOp<IREE::Flow::DispatchTensorLoadOp>(
+        sliceOp, sliceOp.getType(), loadOp.getSource(), loadOp.getSourceDims(),
+        combinedOffsets, combinedSizes, combinedStrides);
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// SwapExtractSliceWithInitTensor
+//===----------------------------------------------------------------------===//
+
+/// Pattern to swap `init_tensor` -> `tensor.extract_slice` with
+/// `init_tensor` of the slice.
+struct SwapExtractSliceWithInitTensor
```

Comment:
Is this still necessary with `SwapExtractSliceWithTiledProducer`? Won't that pull `linalg.init_tensor` in?

Reply:
That method works only for the case where the producer implements the `TilingInterface`. `linalg.init_tensor` doesnt implement that interface. So this is needed for that.